src/generic_factory.h

```diff
@@ -103,23 +104,83 @@ const my_class &string_id<my_class>::obj() const
 
 \endcode
 */
+
+template<typename T>
+class string_id_reader;
+
 template<typename T>
 class generic_factory
 {
     protected:
-        std::unordered_map< string_id<T>, T> data;
+        std::vector<T> list;
+        std::unordered_map<string_id<T>, int_id<T>> map;
 
         std::string type_name;
         std::string id_member_name;
+        std::string alias_member_name;
 
         T &load_override( const string_id<T> &id, JsonObject &jo ) {
             T obj;
+
             obj.id = id;
             obj.load( jo );
             obj.was_loaded = true;
-            T &result = data[id];
-            result = std::move( obj );
-            return result;
+
+            T &inserted_obj = insert( obj );
+
+            if( !alias_member_name.empty() && jo.has_member( alias_member_name ) ) {
+                const int_id<T> i_id = map[id];
+                std::vector<string_id<T>> aliases;
+
+                mandatory( jo, obj.was_loaded, alias_member_name, aliases, string_id_reader<T> {} );
+
+                for( const auto &alias : aliases ) {
+                    const auto iter = map.find( alias );
+
+                    if( iter != map.end() ) {
+                        jo.throw_error( "duplicate " + type_name + " alias \"" + alias.str() + "\" in \"" + id.str() +
+                                        "\"" );
+                    }
+                    map[alias] = i_id;
+                }
+            }
+
+            return inserted_obj;
+        }
+
+        bool find_id( const string_id<T> &id, int_id<T> &result ) const {
+            result = id.get_cid();
+            if( is_valid( result ) && list[result].id == id ) {
+                return true;
+            }
+            const auto iter = map.find( id );
+            if( iter == map.end() ) {
+                return false;
+            }
+            result = iter->second;
+            id.set_cid( result );
+            return true;
+        }
+
+        const T &dummy_obj() const {
+            static const T dummy;
```

Comment:
I tried to do so and (surprisingly) got a segfault right away. Brief investigation revealed, that assignments like `id( NULL_ID )` are sensitive to whether the specific `NULL_ID` was defined above or below the assignment. Specific `NULL_ID`s don't yet exist when in headers (for now), but compiler just doesn't care. It compiles code which fails to run. I changed the code a bit and got rid of the segfault, but it looks like escaping from the issue rather than fixing it. Ideally, broken code should not compile.